addons/product_margin/models/product_product.py

```diff
@@ -128,38 +128,43 @@ def _compute_product_margin_fields_values(self, field_names=None):
                  cr.company_id = i.company_id and
                  cr.date_start <= COALESCE(i.invoice_date, NOW()) and
                  (cr.date_end IS NULL OR cr.date_end > COALESCE(i.invoice_date, NOW())))
-                WHERE l.product_id = %s
+                WHERE l.product_id IN %s
                 AND i.state IN %s
                 AND i.invoice_payment_state IN %s
                 AND i.type IN %s
                 AND i.invoice_date BETWEEN %s AND  %s
                 AND i.company_id = %s
                 AND l.display_type IS NULL
                 AND l.exclude_from_invoice_tab = false
+                GROUP BY l.product_id
                 """.format(self.env['res.currency']._select_companies_rates())
-            invoice_types = ('out_invoice', 'out_refund')
-            self.env.cr.execute(sqlstr, (val.id, states, invoice_payment_states, invoice_types, date_from, date_to, company_id))
-            result = self.env.cr.fetchall()[0]
-            res[val.id]['sale_avg_price'] = result[0] and result[0] or 0.0
-            res[val.id]['sale_num_invoiced'] = result[1] and result[1] or 0.0
-            res[val.id]['turnover'] = result[2] and result[2] or 0.0
-            res[val.id]['sale_expected'] = result[3] and result[3] or 0.0
-            res[val.id]['sales_gap'] = res[val.id]['sale_expected'] - res[val.id]['turnover']
-            ctx = self.env.context.copy()
-            ctx['force_company'] = company_id
-            invoice_types = ('in_invoice', 'in_refund')
-            self.env.cr.execute(sqlstr, (val.id, states, invoice_payment_states, invoice_types, date_from, date_to, company_id))
-            result = self.env.cr.fetchall()[0]
-            res[val.id]['purchase_avg_price'] = result[0] and result[0] or 0.0
-            res[val.id]['purchase_num_invoiced'] = result[1] and result[1] or 0.0
-            res[val.id]['total_cost'] = result[2] and result[2] or 0.0
-            res[val.id]['normal_cost'] = val.standard_price * res[val.id]['purchase_num_invoiced']
-            res[val.id]['purchase_gap'] = res[val.id]['normal_cost'] - res[val.id]['total_cost']
+        invoice_types = ('out_invoice', 'out_refund')
+        self.env.cr.execute(sqlstr, (tuple(self.ids), states, invoice_payment_states, invoice_types, date_from, date_to, company_id))
+        for product_id, avg, qty, total, sale in self.env.cr.fetchall():
+            res[product_id]['sale_avg_price'] = avg and avg or 0.0
+            res[product_id]['sale_num_invoiced'] = qty and qty or 0.0
+            res[product_id]['turnover'] = total and total or 0.0
+            res[product_id]['sale_expected'] = sale and sale or 0.0
+            res[product_id]['sales_gap'] = res[product_id]['sale_expected'] - res[product_id]['turnover']
+            res[product_id]['total_margin'] = res[product_id]['turnover']
```

Comment:
Why is it defined here and it wasn't before? The commit is only talking about perf improvements, but it seems to be doing more than that.